### PR TITLE
[BUGFIX] Add missing exception handling in Search Bar

### LINKED ISSUE #3430: Lack of proper exception handling in Search Bar
Typing in the search bar fetches records (Resources of some kind) and renders them in the either of the following components. 
It seems the last 3 components do not have a proper exception handling which may become problematic. 

```typescript
          <SearchDashboardList query={query} onClick={handleClose} />
           /* Lack of exception handling for the following components!*/
          <SearchProjectList query={query} onClick={handleClose} />
          <SearchGlobalDatasource query={query} onClick={handleClose} />
          <SearchDatasourceList query={query} onClick={handleClose} />
```

### Perses version

> v0.52

### PR DESCRIPTION
**Pull request by ayanarukasar**
# Description

Header search was failing silently for projects, datasources, and global datasources. This change:

- Shows a shared `SearchErrorAlert` when those list APIs fail, with vertically aligned icon and text
- Checks the current user's permissions first and only fetches search data they can read (global datasources require global `read`; project-scoped resources use any matching project permission)
- If a `403` still comes back, shows `you do not have permission to view this` instead of raw RBAC scope/kind names
- Keeps real transport/server errors visible (`Unknown error` when the API message is empty)

Fixes #3430

This supersedes the stale/conflicting #3431. It also addresses review on this PR: skip unauthorized searches rather than only sanitizing the 403 text ([comment from @Nexucis](https://github.com/perses/perses/pull/4344#issuecomment-3618722477), plus [ayanarukasar/perses#1](https://github.com/ayanarukasar/perses/pull/1)).

# Screenshots

Failed list calls (network/5xx) plus a sanitized permission error. Icons are vertically aligned with the alert text:

Search bar error alerts

403 responses do not leak RBAC scope/kind names:

Search bar forbidden alert

# Checklist

- [x] Pull request has a descriptive title and context useful to a reviewer.
- [x] Pull request title follows the `[<catalog_entry>] <commit message>` naming convention using one of the
  following `catalog_entry` values: `FEATURE`, `ENHANCEMENT`, `BUGFIX`, `BREAKINGCHANGE`, `DOC`,`IGNORE`.
- [x] All commits have [DCO signoffs](https://github.com/probot/dco#how-it-works).

## UI Changes

- [x] Changes that impact the UI include screenshots and/or screencasts of the relevant changes.
- [x] Code follows the [UI guidelines](https://github.com/perses/perses/blob/main/ui/ui-guidelines.md).
- [x] E2E tests are stable and unlikely to be flaky.
  See [e2e](https://github.com/perses/perses/tree/main/ui/e2e#visual-tests) docs for more details. Common issues include:
  - Is the data inconsistent? You need to mock API requests.
  - Does the time change? You need to use consistent time values or mock time utilities.
  - Does it have loading states? You need to wait for loading to complete.

## Test plan

- [x] Open Search (Mod+K / click Search)
- [x] With no `GlobalDatasource` read permission, search does not call `/api/v1/globaldatasources` and does not show RBAC internals
- [x] If a list request still returns 403, the alert says "you do not have permission to view this"
- [x] Block `/api/v1/projects` (or equivalent 500) and confirm a "Failed to load projects" alert appears
- [ ] Block `/api/v1/datasources` and confirm a datasources error alert appears
- [ ] Block `/api/v1/globaldatasources` when the user *does* have permission and confirm a generic permission or load error appears
- [ ] Block dashboard list endpoints and confirm the dashboards error alert still appears with aligned icon/text
- [x] With all APIs healthy, search still lists matching resources as before

### CODE REVIEW

**Comment by jgbernalp** [comment]
@ayanarukasar please follow the PR template, even when using AI tools. If UI changes are made they should include an image, the test plan has just empty checkboxes.

**Comment by Sarthak-Shreshtha01** [comment]
SearchErrorAlert's detail falls back to the raw StatusError.message, which for a 403 shows internal RBAC scope/kind names directly to the user, e.g. forbidden access: missing 'read' global permission for 'GlobalDatasource' kind. That's unactionable for a normal user and exposes backend permission-model internals.

Opened a small patch on top of this branch to fix it: https://github.com/ayanarukasar/perses/pull/1

@jgbernalp since you've been reviewing this PR — what's the preferred way to handle this here? 
Happy to wait for @ayanarukasar to pick up the patch, or open a PR directly against perses/perses if that's easier to review. 
Wanted to check before assuming either way.

**Comment by Nexucis** [comment]
@ayanarukasar would you like to continue this PR ? If yes can you resolve the conflict and address the comment from @jgbernalp ? Thank you

> SearchErrorAlert's detail falls back to the raw StatusError.message, which for a 403 shows internal RBAC scope/kind names directly to the user, e.g. forbidden access: missing 'read' global permission for 'GlobalDatasource' kind. That's unactionable for a normal user and exposes backend permission-model internals.
> 
> Opened a small patch on top of this branch to fix it: [ayanarukasar#1](https://github.com/ayanarukasar/perses/pull/1)
> 
> @jgbernalp since you've been reviewing this PR — what's the preferred way to handle this here? Happy to wait for @ayanarukasar to pick up the patch, or open a PR directly against perses/perses if that's easier to review. Wanted to check before assuming either way.

I think the best way to tackle this issue is actually to avoid searching through the data you don't have access too. And for that we need to check the user permission and then based on that to get the data he has accessed to

**Comment by ayanarukasar** [comment]
Thanks for the ping — I continued this PR.

- Rebased/merged `main` and resolved the `SearchBar.tsx` conflict.
- Followed the PR template (description, checklist, test plan).
- @Nexucis: search now checks permissions first and only fetches resources the user can read (`useHasPermission` for global datasources, `useHasPermissionInAnyProject` for project-scoped lists). Unauthorized types are skipped instead of being searched and then failing with a 403.
- @Sarthak-Shreshtha01: if a 403 still happens, `SearchErrorAlert` shows a generic "you do not have permission to view this" and does not render raw RBAC scope/kind names. Your patch is covered; we went a step further by not issuing the unauthorized list calls.
- @jgbernalp: template/checklist filled. I could not attach a live UI screenshot from this environment because the API backend was not running; unit tests cover the error copy, alignment styles, and the permission-gated fetch. I can add a local screenshot if you still want one on the PR.

**Comment by ayanarukasar** [comment]
Addressed the Copilot review:

- **Project search for Kubernetes users:** search no longer requires a `Project` scope. `useCanReadAnyProject` matches backend `checkNamespaceAccess` — native `Project` read, or Kubernetes/delegated read on `Dashboard`, `Datasource`, or `Secret`.
- **Permissions query failures:** loading and error are no longer treated as an empty permission map. While permissions are loading, search does not show “No records found.” If `/users/{name}/permissions` fails, SearchBar shows `Failed to load permissions` instead of an empty result.

**Comment by ayanarukasar** [comment]
Addressed the latest Copilot review:

- **Native vs Kubernetes project gates:** `useCanReadAnyProject` now uses the `Project` scope for native RBAC. Dashboard/Datasource/Secret fallbacks apply only when delegated/Kubernetes auth is enabled.
- **Stale search flags:** `useSearchListErrorState` clears `hasResource` and `hasLoadError` when a gated list unmounts, so “No records found” can show after access is revoked.

Also merged latest `main`.

File: .pr-screenshots/search-bar-error-alerts.png
Comment (Gladorme) [inline .pr-screenshots/search-bar-error-alerts.png:1]: Please ignore this folder, we don't want image in this repo

File: ui/app/src/components/Header/SearchBar/SearchBar.tsx [hunk 1]
```diff
@@ -43,18 +50,76 @@ interface ResourceListProps {
...
+  );
+}
+
+function useSearchListErrorState(
+  type: ResourceType,
+  error: StatusError | null | undefined,
+  isResources?: (type: ResourceType, available: boolean) => void,
+  onLoadError?: (type: ResourceType, hasError: boolean) => void,
+): StatusError | undefined {
+  useEffect(() => {
```
Comment (Gladorme) [inline ui/app/src/components/Header/SearchBar/SearchBar.tsx]: Not sure what is done here, but pretty sure it's is a misusage of useEffect.

File: ui/app/src/components/Header/SearchBar/SearchBar.tsx [hunk 2]
```diff
@@ -64,15 +129,28 @@ function SearchProjectList(props: ResourceListProps): ReactElement | null {
...
-  const { query, onClick, isResources } = props;
+  const { data: globalDatasourceList, error: globalDatasourceListError } = useGlobalDatasourceList({
+    refetchOnMount: false,
+  });
+  const { query, onClick, isResources, onLoadError } = props;
   const handleIsResource = useCallback(
     (isAvailable: boolean): void => isResources?.('globalDatasources', isAvailable),
     [isResources],
   );
+  const visibleError = useSearchListErrorState(
```
Comment (Gladorme) [inline ui/app/src/components/Header/SearchBar/SearchBar.tsx]: To me, `useSearchListErrorState` should be called in `SearchErrorAlert`

File: ui/app/src/context/Authorization.tsx [hunk 3]
```diff
@@ -24,10 +24,18 @@ import { useIsDelegatedAuthnProviderEnabled, useIsAuthEnabled } from './Config';
...
+// access is derived from read on Dashboard, Datasource, or Secret in a namespace
+// (see internal/api/authorization/k8s/k8s.go checkNamespaceAccess).
+const DELEGATED_PROJECT_READ_SCOPES: Scope[] = ['Project', 'Dashboard', 'Datasource', 'Secret'];
+
 interface AuthorizationContext {
   enabled: boolean;
   username: string;
   userPermissions: Record<string, Permission[]>;
+  isPermissionsLoading: boolean;
+  permissionsError: StatusError | undefined;
```
Comment (Gladorme) [inline ui/app/src/context/Authorization.tsx]: I don't think we want that here.

**Review by Gladorme — COMMENTED** [review]
I did a first quick review, but I need to review with more time. Because some stuff looks strange / not correct to me 😅

**Comment by ayanarukasar** [comment]
Thanks @Gladorme — addressed the first-pass comments:

- Removed `.pr-screenshots/` from the repo. The PR description now describes the error copy instead of embedding images.
- `useSearchListErrorState` is only called from `SearchErrorAlert`, so the effect runs when that error UI mounts/unmounts rather than from every list hook.
- Permission query loading/error is no longer stored on `AuthorizationContext`. `usePermissionsQueryStatus` reads the permissions query directly.

File: internal/api/e2e/framework/server.go [hunk 4]
```diff
@@ -159,10 +159,27 @@ func CreateAuthorizationHeader(token string) (string, string) {
 	return "Authorization", fmt.Sprintf("Bearer %s", token)
 }
 
+func hasExtractedPlugins(pluginPath string) bool {
```
Comment (Gladorme) [inline internal/api/e2e/framework/server.go]: Not related to the PR, need to be in its own PR

File: ui/app/src/components/Header/SearchBar/SearchBar.tsx [hunk 5]
```diff
@@ -43,18 +50,90 @@ interface ResourceListProps {
...
+    onLoadError?.(type, true);
+    isResources?.(type, false);
+    return (): void => {
+      onLoadError?.(type, false);
+      isResources?.(type, false);
+    };
+  }, [isResources, onLoadError, type]);
+}
+
+function SearchErrorAlert({ title, error, type, isResources, onLoadError }: SearchErrorAlertProps): ReactElement {
```
Comment (Gladorme) [inline ui/app/src/components/Header/SearchBar/SearchBar.tsx]: Create a dedicated file for this component

File: ui/app/src/components/Header/SearchBar/SearchBar.tsx [hunk 6]
```diff
@@ -43,18 +50,90 @@ interface ResourceListProps {
...
+  isResources?: (type: ResourceType, available: boolean) => void;
+  onLoadError?: (type: ResourceType, hasError: boolean) => void;
+}
+
+function useSearchListErrorState(
+  type: ResourceType | undefined,
+  isResources?: (type: ResourceType, available: boolean) => void,
+  onLoadError?: (type: ResourceType, hasError: boolean) => void,
+): void {
+  useEffect(() => {
```
Comment (Gladorme) [inline ui/app/src/components/Header/SearchBar/SearchBar.tsx]: Misusage of `useEffect`. UseEffect is made to connect to external system (https://react.dev/reference/react/useEffect#usage). From what I understood from your code, this is "internal" data from children to parent node: `useEffect` is not made for that.

**Comment by ayanarukasar** [comment]
Thanks @Gladorme — addressed the latest comments:

- Dropped the e2e plugin unzip change from this PR (`server.go` / `install_plugin.go`). That belongs in a follow-up.
- `SearchErrorAlert` now lives in its own file.
- Removed the child-to-parent `useEffect`. SearchBar owns the list queries (with `enabled` gates) and derives load errors from those results instead of callbacks from the alert.

**Comment by Gladorme** [comment]
Loading a dashboard is now fetching all dashboard metadata, not great. Before it was only fetched once the searchbox was opened. 

The code is working, but I have a weird feeling when reviewing the code. I will try to take a look on my side to see if it can be implemented differently.
